alpheios_nemo_ui/__init__.py

```diff
@@ -534,6 +535,147 @@ def new_level(self,subreference,prev):
+            # which could be ranges so we want just the end of the previous and the
+            # beginning of the next
+            if passage.prevId:
+                prevPassageRef = Reference(passage.prevId)
+                prevPassage = prevPassageRef.start
+                if prevPassageRef.end is not None:
+                    prevPassage = prevPassageRef.end
+            else:
+                prevPassage = None
+            if passage.nextId:
+                nextPassage = Reference(passage.nextId).start
+            else:
+                nextPassage = None
+
+            # turn our subreference string into a real Reference object and
+            # find its start and end (which will be the same if it was a single
+            # reference and not a range)
+            subreff = Reference(subreference)
+            substart = subreff.start
+            subend = subreff.end
+            if subend is None:
+                subend = substart
+
+            # now loop through the precaculated references and try to find one that
+            # the custom reference is part of
+            for index,reff in enumerate(reffs):
+                reff = Reference(reff)
+                refstart = reff.start
+                refend = reff.end
+                # these should all be ranges but if not just make the end the start
+                if refend is None:
+                    refend = reff.start
+                if refstart == substart:
+                    # custom reference starts the exact beginning of
+                    # a precalculated reference
+                    # so the previous precalculated reference is retained whole
+                    # and the next reference is the immediate next sibling to the
+                    #  end of the precalculated range
+                    nextRange = str(nextPassage) + '-' + str(refend)
+                    if index > 0:
+                        prevRange = reffs[index-1]
+                    else:
+                        prevRange = None
+                    return prevRange, nextRange
+                elif refend == subend:
+                    # custom reference ends at the exact end of the precalculated
+                    # reference so the previous reference is the beginning
+                    # of the precalculated reference to the previous sibling and
+                    # the next reference is retained whole
+                    prevRange = str(refstart) + "-" + str(prevPassage)
+                    if index < len(reffs) - 1:
+                        nextRange = reffs[index+1]
+                    else:
+                        nextRange = None
+                    return prevRange, nextRange
+                else:
+                    # check to see if the custom reference is somewhere in the middle
+                    latest_start = self.latest_reff(refstart, substart)
+                    latest_end = self.latest_reff(refend,subend)
```

Comment:
the latest_reff function tries to compare 2 references and decide based upon the numeric values of the parts, which might come later in the text. 
What I want is to find where the subreference entered by the user falls within the predefined range. so the test following this line is true of the subreference start is later than the precalculated reference start and the subreference end is earlier than the precalculated reference end. 

So you are probably right that it is a little confusing to use the latest_reff test for each, and I could have an earliest_reff function which did the opposite of latest_reff but it would just be a reverse of the logic, so instead I just reversed the test. 

not sure if that explanation makes it more or less confusing though :-)

Reply:
Thanks for the explanation, it's much more understandable now to me